hopsanc/src/hopsanc.cpp

```diff
@@ -0,0 +1,275 @@
+{
+    if(silent) return;
+
+    hopsan::HString msg, type, tag;
+    while (hopsanCore.checkMessage() > 0) {
+        hopsanCore.getMessage(msg,type,tag);
+        if (type == "debug") {
+            if (printDebug) {
+                std::cout << msg.c_str() << "\n";
+            }
+        }
+        else {
+            std::cout << msg.c_str() << "\n";
+        }
+    }
+}
+
+
+int loadModel(const char* path) {
+    spCoreComponentSystem = gHopsanCore.loadHMFModelFile(path, startTime, stopTime);
+    if(!spCoreComponentSystem) {
+        std::cout << "Failed to instantiate model!\n";
+        printWaitingMessages(gHopsanCore, false, false);
+        return -1;
+    }
+    const hopsan::HString modelName = spCoreComponentSystem->getName();
+    spCoreComponentSystem->addSearchPath(modelName+"-resources");
+    std::cout << "Loaded model: " << modelName.c_str() << "\n";
+    printWaitingMessages(gHopsanCore, false, false);
+    return 0;
+}
+
+
+//! @brief Returns specified data vector as C array
+//! @param [in] variable Variable name ("component.port.variable")
+//! @param [out] size Size of returned data array
+//! @returns Pointer to data array
+double *getDataVector(const char* variable, int &size)
+{
+    if(!spCoreComponentSystem) {
+        std::cout << "Error: No model is loaded.\n";
+        size = -1;
+        return nullptr;
+    }
+
+    //Parse variable string
+    hopsan::HString varStr(variable);
+    hopsan::HVector<hopsan::HString> splitSys = varStr.split('|');
+    hopsan::HVector<hopsan::HString> splitVar = splitSys.last().split('.');
+    splitSys.resize(splitSys.size()-1);
+    if(splitVar.size() < 3) {
+        std::cout << "Error: Component name, port name and variable name must be specified.\n";
+        size = -1;
+        return nullptr;
+    }
+
+    //Find system
+    hopsan::ComponentSystem *pSystem = spCoreComponentSystem;
+    for(size_t i=0; i<splitSys.size(); ++i) {
+        pSystem = pSystem->getSubComponentSystem(splitSys[i]);
+        if(!pSystem) {
+            std::cout << "Error: Subsystem not found: " << splitSys[i].c_str() << "\n";
+            size = -1;
+            return nullptr;
+        }
+    }
+
+    //Find component
+    hopsan::Component *pComp = pSystem->getSubComponent(splitVar[0]);
+    if(!pComp) {
+        std::cout << "Error: No such component: " << splitVar[0].c_str() << "\n";
+        size = -1;
+        return nullptr;
+    }
+
+    //Find port
+    hopsan::Port *pPort = pComp->getPort(splitVar[1]);
+    if(!pPort) {
+        std::cout << "Error: No such port: " << splitVar[1].c_str() << "\n";
+        size = -1;
+        return nullptr;
+    }
+
+    int varId = pPort->getNodeDataIdFromName(splitVar[2]);
+    if(varId < 0) {
+        std::cout << "Error: No such variable: " << splitVar[2].c_str() << "\n";
+        size = -1;
+        return nullptr;
+    }
+
+    size = int(spCoreComponentSystem->getNumActuallyLoggedSamples());
+    double *data = new double[size];
```

Comment:
Returning this array will lead to a memory leak since it is never deleted, also it can not be deleted in the caller (at least on Windows) since that is in a different heap, which will lead to memory corruption.

One alternative would be to return the array to this library for deletion. 
But I think that it would be better if the caller allocates the data buffer and sends in a pointer to where the data should be written. Of course the caller must then fist know the length of the data, but a function call for that information can be implemented.

Then the caller calls this function with a pointer to a double array, and the maximum allowed size.

Reply:
I changed it according to your second suggestion. It actually made the Julia code much nicer as well :)